File: client.ts/src/implementations/evm.ts
import { JsonRpcProvider, Provider } from "@ethersproject/providers";
import { Wallet } from "@ethersproject/wallet";
import type { Signer } from "@ethersproject/abstract-signer";

import * as typechain from "hapi-core-evm/typechain-types";

import {
  Addr,
  Address,
  Amount,
  Asset,
  Case,
  CaseStatus,
  Category,
  HapiCore,
  HapiCoreAddresses,
  HapiCoreNetwork,
  Reporter,
  ReporterRole,
  Result,
  RewardConfiguration,
  StakeConfiguration,
  Uuid,
} from "../interface";
import { bigIntToUuid, uuidToBigNumberish } from "../util";
import { BigNumber } from "@ethersproject/bignumber";

export interface EvmConnectionOptions {
  network: HapiCoreNetwork.Ethereum | HapiCoreNetwork.BSC;
  provider: Provider | EvmProviderOptions;
  address?: Addr;
  signer?: Signer;
  signerPrivateKey?: string;
}

export interface EvmProviderOptions {
  providerUrl: string;
}

export class HapiCoreEvm implements HapiCore {
  private contract: typechain.HapiCore;

  constructor(options: EvmConnectionOptions) {
    if (options.provider.constructor.name !== "Provider") {
      options.provider = new JsonRpcProvider(
        (options.provider as EvmProviderOptions).providerUrl
      );
    }

    if (options.signerPrivateKey && !options.signer) {
      options.signer = new Wallet(
        options.signerPrivateKey,
        options.provider as Provider
      );
    }

    this.contract = typechain.HapiCore__factory.connect(
      options.address || HapiCoreAddresses[options.network],
      options.signer || (options.provider as Provider)
    );
  }

  async setAuthority(address: Addr): Promise<Result> {
    const tx = await this.contract.setAuthority(address);
    return { transactionHash: tx.hash };
  }

  async getAuthority(): Promise<Addr> {
    return await this.contract.authority();
  }

  async updateStakeConfiguration(
    token: Addr,
    unlockDuration: number,
    validatorStake: Amount,
    tracerStake: Amount,
    publisherStake: Amount,
    authorityStake: Amount
  ): Promise<Result> {
    const tx = await this.contract.updateStakeConfiguration(
      token,
      unlockDuration,
      validatorStake,
      tracerStake,
      publisherStake,
      authorityStake
    );
    return { transactionHash: tx.hash };
  }

  async getStakeConfiguration(): Promise<StakeConfiguration> {
    const stakeConfiguration = await this.contract.stakeConfiguration();
    return {
      token: stakeConfiguration.token,
      unlockDuration: stakeConfiguration.unlock_duration.toNumber(),
      validatorStake: stakeConfiguration.validator_stake.toString(),
      tracerStake: stakeConfiguration.tracer_stake.toString(),
      publisherStake: stakeConfiguration.publisher_stake.toString(),
      authorityStake: stakeConfiguration.authority_stake.toString(),
    };
  }

  async updateRewardConfiguration(
    token: Addr,
    addressConfirmationReward: Amount,
    traceReward: Amount
  ): Promise<Result> {
    const tx = await this.contract.updateRewardConfiguration(
      token,
      addressConfirmationReward,
      traceReward
    );
    return { transactionHash: tx.hash };
  }

  async getRewardConfiguration(): Promise<RewardConfiguration> {
    const rewardConfiguration = await this.contract.rewardConfiguration();
    return {
      token: rewardConfiguration.token,
      addressConfirmationReward:
        rewardConfiguration.address_confirmation_reward.toString(),
      tracerReward: rewardConfiguration.tracer_reward.toString(),
    };
  }

  async createReporter(
    id: Uuid,
    role: ReporterRole,
    account: Addr,
    name: string,
    url: string
  ): Promise<Result> {
    const tx = await this.contract.createReporter(
      uuidToBigNumberish(id),
      account,
      role.toString(),
      name,
      url
    );

    return { transactionHash: tx.hash };
  }

  async getReporter(id: Uuid): Promise<Reporter> {
    const reporter = await this.contract.getReporter(uuidToBigNumberish(id));
    return {
      id: bigIntToUuid(reporter.id),
      account: reporter.account,
      role: reporter.role,
      status: reporter.status,
      name: reporter.name,
      url: reporter.url,
      stake: reporter.stake.toString(),
      unlockTimestamp: reporter.unlock_timestamp.toNumber(),
    };
  }

  async updateReporter(
    id: Uuid,
    role: ReporterRole,
    account: Addr,
    name: string,
    url: string
  ): Promise<Result> {
    const tx = await this.contract.updateReporter(
      uuidToBigNumberish(id),
      role.toString(),
      account,
      name,
      url
    );
    return { transactionHash: tx.hash };
  }

  async activateReporter(): Promise<Result> {
    const tx = await this.contract.activateReporter();
    return { transactionHash: tx.hash };
  }

  async deactivateReporter(): Promise<Result> {
    const tx = await this.contract.deactivateReporter();
    return { transactionHash: tx.hash };
  }

  async unstakeReporter(): Promise<Result> {
    const tx = await this.contract.unstake();
    return { transactionHash: tx.hash };
  }

  async createCase(id: string, name: string, url: string): Promise<Result> {
    const tx = await this.contract.createCase(
      uuidToBigNumberish(id),
      name,
      url
    );
    return { transactionHash: tx.hash };
  }

  async getCase(id: Uuid): Promise<Case> {
    const c = await this.contract.getCase(uuidToBigNumberish(id));
    return {
      id: bigIntToUuid(c.id),
      name: c.name,
      url: c.url,
      status: c.status,
    };
  }

  async updateCase(
    id: Uuid,
    name: Addr,
    url: string,
    status: CaseStatus
  ): Promise<Result> {
    const tx = await this.contract.updateCase(
      uuidToBigNumberish(id),
      name,
      url,
      status.toString()
    );
    return { transactionHash: tx.hash };
  }

  async createAddress(
    address: Addr,
    caseId: Uuid,
    risk: number,
    category: Category
  ): Promise<Result> {
    const tx = await this.contract.createAddress(
      address,
      uuidToBigNumberish(caseId),
      risk,
      category.toString()
    );
    return { transactionHash: tx.hash };
  }

  async getAddress(address: Addr): Promise<Address> {
    const a = await this.contract.getAddress(address);
    return {
      address: a.addr.toString(),
      caseId: bigIntToUuid(a.case_id),
      reporterId: a.reporter_id.toString(),
      risk: a.risk,
      category: a.category,
    };
  }

  async updateAddress(
    address: Addr,
    caseId: Uuid,
    risk: number,
    category: Category
  ): Promise<Result> {
    const tx = await this.contract.updateAddress(
      address,
      uuidToBigNumberish(caseId),
      risk,
      category.toString()
    );
    return { transactionHash: tx.hash };
  }

  async createAsset(
    address: Addr,
    assetId: string,
    caseId: Uuid,
    risk: number,
    category: Category
  ): Promise<Result> {
    const tx = await this.contract.createAsset(
      address,
      assetId,
      uuidToBigNumberish(caseId),
      risk,
      category.toString()
    );
    return { transactionHash: tx.hash };
  }

  async getAsset(address: Addr, assetId: string): Promise<Asset> {
    const a = await this.contract.getAsset(address, assetId);
    return {
      address: a.addr.toString(),
      assetId: a.asset_id.toString(),
      caseId: bigIntToUuid(a.case_id),
      reporterId: a.reporter_id.toString(),
      risk: a.risk,
      category: a.category,
    };
  }

  async updateAsset(
    address: Addr,
    assetId: string,
    caseId: Uuid,
    risk: number,
    category: Category
  ): Promise<Result> {
    const tx = await this.contract.updateAsset(
      address,
      assetId,
      caseId,
      risk,
      category.toString()
    );
    return { transactionHash: tx.hash };
  }
}

File: client.rs/src/hapi_core/implementations/evm.rs
use std::{str::FromStr, sync::Arc};

use async_trait::async_trait;
use ethers::{
    prelude::{abigen, SignerMiddleware},
    providers::{Http, Provider},
    signers::LocalWallet,
    types::Address as EthAddress,
};
use ethers_providers::Middleware;
use ethers_signers::Signer;

use crate::{
    hapi_core::{
        address::{Address, CreateAddressInput, UpdateAddressInput},
        asset::{Asset, AssetId, CreateAssetInput, UpdateAssetInput},
        case::{Case, CreateCaseInput, UpdateCaseInput},
        configuration::{RewardConfiguration, StakeConfiguration},
        reporter::{CreateReporterInput, Reporter, UpdateReporterInput},
        result::{ClientError, Result, Tx},
    },
    HapiCore,
};

pub struct HapiCoreEvmOptions {
    pub provider_url: String,
    pub contract_address: String,
    pub private_key: Option<String>,
}

abigen!(HAPI_CORE, "./src/hapi_core/implementations/abi.json");

pub struct HapiCoreEvm {
    provider: Provider<Http>,
    signer: LocalWallet,
    contract: HAPI_CORE<SignerMiddleware<Provider<Http>, LocalWallet>>,
    client: Arc<SignerMiddleware<Provider<Http>, LocalWallet>>,
}

impl HapiCoreEvm {
    pub fn new(options: HapiCoreEvmOptions) -> Result<Self> {
        let provider = Provider::<Http>::try_from(options.provider_url.as_str())
            .map_err(|e| ClientError::UrlParseError(e.to_string()))?;

        let signer = LocalWallet::from_str(options.private_key.unwrap_or_default().as_str())
            .map_err(|e| ClientError::Ethers(e.to_string()))?
            .with_chain_id(31337_u64);

        let client = SignerMiddleware::new(provider.clone(), signer.clone());

        let client = Arc::new(client);

        let contract_address: EthAddress = options.contract_address.parse().map_err(
            |e: <EthAddress as std::str::FromStr>::Err| ClientError::EthAddressParse(e.to_string()),
        )?;

        let contract: HAPI_CORE<SignerMiddleware<Provider<Http>, LocalWallet>> =
            HAPI_CORE::new(contract_address, client.clone());

        Ok(Self {
            provider,
            signer,
            contract,
            client,
        })
    }
}

#[async_trait]
impl HapiCore for HapiCoreEvm {
    async fn set_authority(&self, address: &str) -> Result<Tx> {
        let authority: EthAddress =
            address
                .parse()
                .map_err(|e: <EthAddress as std::str::FromStr>::Err| {
                    ClientError::EthAddressParse(e.to_string())
                })?;

        let tx = self
            .contract
            .set_authority(authority)
            .send()
            .await
            .map_err(|e| ClientError::Ethers(format!("set_authority failed: {e}")))?
            .await?;

        if let Some(receipt) = tx {
            Ok(Tx {
                hash: format!("{:?}", receipt.transaction_hash),
            })
        } else {
            Err(ClientError::Ethers(
                "set_authority failed: no receipt".to_string(),
            ))
        }
    }

    async fn get_authority(&self) -> Result<String> {
        self.contract
            .authority()
            .call()
            .await
            .map_err(|e| ClientError::Ethers(format!("get_authority failed: {e}")))
            .map(|a| format!("{a:?}"))
    }

    async fn update_stake_configuration(&self, _configuration: StakeConfiguration) -> Result<Tx> {
        unimplemented!()
    }
    async fn get_stake_configuration(&self) -> Result<StakeConfiguration> {
        unimplemented!()
    }

    async fn update_reward_configuration(&self, _configuration: RewardConfiguration) -> Result<Tx> {
        unimplemented!()
    }
    async fn get_reward_configuration(&self) -> Result<RewardConfiguration> {
        unimplemented!()
    }

    async fn create_reporter(&self, _input: CreateReporterInput) -> Result<Tx> {
        unimplemented!()
    }
    async fn update_reporter(&self, _input: UpdateReporterInput) -> Result<Tx> {
        unimplemented!()
    }
    async fn get_reporter(&self, _id: &str) -> Result<Reporter> {
        unimplemented!()
    }
    async fn get_reporter_count(&self) -> Result<u64> {
        unimplemented!()
    }
    async fn get_reporters(&self, _skip: u64, _take: u64) -> Result<Vec<Reporter>> {
        unimplemented!()
    }

    async fn activate_reporter(&self) -> Result<Tx> {
        unimplemented!()
    }
    async fn deactivate_reporter(&self) -> Result<Tx> {
        unimplemented!()
    }
    async fn unstake_reporter(&self) -> Result<Tx> {
        unimplemented!()
    }

    async fn create_case(&self, _input: CreateCaseInput) -> Result<Tx> {
        unimplemented!()
    }
    async fn update_case(&self, _input: UpdateCaseInput) -> Result<Tx> {
        unimplemented!()
    }
    async fn get_case(&self, _id: &str) -> Result<Case> {
        unimplemented!()
    }
    async fn get_case_count(&self) -> Result<u64> {
        unimplemented!()
    }
    async fn get_cases(&self, _skip: u64, _take: u64) -> Result<Vec<Case>> {
        unimplemented!()
    }

    async fn create_address(&self, _input: CreateAddressInput) -> Result<Tx> {
        unimplemented!()
    }
    async fn update_address(&self, _input: UpdateAddressInput) -> Result<Tx> {
        unimplemented!()
    }
    async fn get_address(&self, _addr: &str) -> Result<Address> {
        unimplemented!()
    }
    async fn get_address_count(&self) -> Result<u64> {
        unimplemented!()
    }
    async fn get_addresses(&self, _skip: u64, _take: u64) -> Result<Vec<Address>> {
        unimplemented!()
    }

    async fn create_asset(&self, _input: CreateAssetInput) -> Result<Tx> {
        unimplemented!()
    }
    async fn update_asset(&self, _input: UpdateAssetInput) -> Result<Tx> {
        unimplemented!()
    }
    async fn get_asset(&self, _address: &str, _id: &AssetId) -> Result<Asset> {
        unimplemented!()
    }
    async fn get_asset_count(&self) -> Result<u64> {
        unimplemented!()
    }
    async fn get_assets(&self, _skip: u64, _take: u64) -> Result<Vec<Asset>> {
        unimplemented!()
    }
}
